ring/replication_strategy.go

```diff
@@ -65,25 +77,30 @@ func (s *defaultReplicationStrategy) Filter(instances []InstanceDesc, op Operati
 	return instances, len(instances) - minSuccess, nil
 }
 
-type ignoreUnhealthyInstancesReplicationStrategy struct{}
+type ignoreUnhealthyInstancesReplicationStrategy struct {
+	logger log.Logger
+}
 
-func NewIgnoreUnhealthyInstancesReplicationStrategy() ReplicationStrategy {
-	return &ignoreUnhealthyInstancesReplicationStrategy{}
+func NewIgnoreUnhealthyInstancesReplicationStrategy(logger log.Logger) ReplicationStrategy {
+	return &ignoreUnhealthyInstancesReplicationStrategy{logger: logger}
 }
 
 func (r *ignoreUnhealthyInstancesReplicationStrategy) Filter(instances []InstanceDesc, op Operation, _ int, heartbeatTimeout time.Duration, _ bool) (healthy []InstanceDesc, maxFailures int, err error) {
 	now := time.Now()
 	// Filter out unhealthy instances.
+	var skipped []string
 	for i := 0; i < len(instances); {
 		if instances[i].IsHealthy(op, heartbeatTimeout, now) {
 			i++
 		} else {
+			skipped = append(skipped, instances[i].Addr)
 			instances = append(instances[:i], instances[i+1:]...)
 		}
 	}
 
 	// We need at least 1 healthy instance no matter what is the replication factor set to.
 	if len(instances) == 0 {
+		level.Error(r.logger).Log("msg", "failed to find any healthy ring replicas", "unhealthy", skipped)
```

Comment:
The `Filter` function is used on the write path too and this function could be called at a very high rate. The risk of flooding logs because of this is high, so I'm not sure about it.

I would rather add the unhealthy replicas to the returned error, instead of logging it.

_Comment applies to `defaultReplicationStrategy` too (actually that one is used on write path, but I would keep the same approach here too)._

Reply:
@pracucci did we see all these errors getting logged when we were investigating the ruler/ring failures? Guess we should make sure they do get logged for relevant scenarios. We can sync on this tomorrow.

Reply:
Yes, they're logged by ruler.